[feat] Add additional CFI checks to firmware image library

mhatrevi opened the pull request:
This change addresses item 1 from the issue# https://github.com/chipsalliance/caliptra-sw/issues/982

mhatrevi committed:
[feat] Add additional CFI checks to firmware image library

This change addresses item 1 from the issue# https://github.com/chipsalliance/caliptra-sw/issues/982

korran commented on image/verify/src/verifier.rs:
```diff
@@ -216,24 +221,32 @@ impl<Env: ImageVerificationEnv> ImageVerifier<Env> {
+                if cfi_launder(revocation.contains(key)) {
                     Err(CaliptraError::IMAGE_VERIFIER_ERR_VENDOR_ECC_PUB_KEY_REVOKED)?;
+                } else {
+                    cfi_assert!(!revocation.contains(key));
```
Need to double check the assembly that this re-does the full contains operations instead of just recomparing the result (which is what I suspect).

mhatrevi replied:
Yeah, revocation should be laundered. Hardened it even further by laundering the key as well.

![image](https://github.com/chipsalliance/caliptra-sw/assets/38512878/86776926-0c7c-407d-8d84-ca81b6274915)

korran commented on fmc/tests/fmc_integration_tests/test_hand_off.rs:
```diff
@@ -4,7 +4,7 @@ use caliptra_hw_model::{BootParams, HwModel, InitParams};
 #[test]
 fn test_hand_off() {
-    let rom = caliptra_builder::build_firmware_rom(&firmware::ROM_WITH_UART).unwrap();
+    let rom = caliptra_builder::build_firmware_rom(&firmware::ROM_WITH_UART_NO_CFI).unwrap();
```
revert this change

korran committed:
Add #[inline(never)] to some CFI-wrapped functions (#1053)

These functions were being called multiple times in the ROM, and the
compiler heuristics were bloating the code by inlining them when it
shouldn't have.

Also, make sure that the cfi derive passes explicit inline attributes to
the inner function.

mhatrevi committed:
Fixing tests.

mhatrevi committed:
Merge branch 'main' into vmhatre/verifier-cfi

mhatrevi committed:
Updating ROM hash

korran commented on test/tests/caliptra_integration_tests/smoke_test.rs:
```diff
@@ -541,7 +541,7 @@ fn fips_cmd_test_rt() {
         .set_device_lifecycle(DeviceLifecycle::Unprovisioned);
 
-    let rom = caliptra_builder::build_firmware_rom(&firmware::ROM_WITH_UART).unwrap();
+    let rom = caliptra_builder::build_firmware_rom(&firmware::ROM_WITH_UART_NO_CFI).unwrap();
```
revert this?

korran commented on builder/src/firmware.rs:
```diff
@@ -17,6 +17,12 @@ pub const ROM_WITH_UART: FwId = FwId {
     features: &["emu"],
 };
 
+pub const ROM_WITH_UART_NO_CFI: FwId = FwId {
```
Either remove this or add it to REGISTERED_FW

mhatrevi committed:
Disabling CFI for verifier lib in Runtime

korran commented on image/verify/src/verifier.rs:
```diff
@@ -216,24 +226,32 @@ impl<Env: ImageVerificationEnv> ImageVerifier<Env> {
                     CaliptraError::IMAGE_VERIFIER_ERR_UPDATE_RESET_VENDOR_ECC_PUB_KEY_IDX_MISMATCH,
                 )?;
+            } else {
+                cfi_assert_eq(expected, key_idx);
```
We should be going back to the source here:

```
cfi_assert_eq(self.env.vendor_ecc_pub_key_idx_dv(), key_idx);
```

korran commented on image/verify/src/verifier.rs:
```diff
@@ -253,23 +271,31 @@ impl<Env: ImageVerificationEnv> ImageVerifier<Env> {
                     CaliptraError::IMAGE_VERIFIER_ERR_UPDATE_RESET_VENDOR_LMS_PUB_KEY_IDX_MISMATCH,
                 )?;
+            } else {
+                cfi_assert_eq(expected, key_idx);
```
Probably worth going back to the source here:

```
cfi_assert_eq(self.env.vendor_lms_pub_key_idx_dv(), key_idx);
```

mhatrevi committed:
Addressing feedback

korran commented on image/verify/src/verifier.rs:
```diff
@@ -760,15 +835,18 @@ impl<Env: ImageVerificationEnv> ImageVerifier<Env> {
 }
 
 #[cfg(all(test, target_family = "unix"))]
+#[cfg(feature = "no-cfi")]
```
This is dangerous. With this change, `cargo test` no longer runs these tests.

korran replied:
I'm not sure we have a lot of options to fix this. The first one that comes to mind is to define CFI_STATE_ORG and cfi_panic_handler

mhatrevi committed:
Updating ROM hash

korran committed:
Fix standalone crate test failures.

This makes `cargo test -p caliptra-image-verify` and `cargo test -p
caliptra-runtime` pass.
